### PR TITLE
Vectorize rrect_blur

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/148496

28% speed improvement in the rrect_blur shader.

## before
Screenshot 2024-10-01 at 3 45 46 PM

## after
Screenshot 2024-10-01 at 3 42 30 PM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jtmcdole — APPROVED** [review]

**Comment by chinmaygarde** [comment]
Really curious to see the malioc diff.

**Comment by gaaclarke** [comment]
> Really curious to see the malioc diff.

Here's one from before my last (and final) vectorization: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8735240889996581441/+/u/test:_malioc_diff/stdout

**Comment by gaaclarke** [comment]
One thing that is unfortunate is that it seems to falling into "stack spilling".  I would have expected less or more register usage.  It possibly could be faster if we could avoid that.

**Comment by gaaclarke** [comment]
> It's faster on iOS because there are likely no spills.

Or there is spilling on iOS too but the savings from the vectorization are larger than the load/store calls.  I could try to reuse variables more to see if that reduces register pressure but I can't run malioc locally.  Do we have an easy way to do that?  I have it installed locally but it doesn't handle includes and doesn't seem to have an argument to make them work.

```
$ /Applications/Arm_Performance_Studio_2024.3/mali_offline_compiler/malioc --opengles --fragment --core Mali-G78 impeller/entity/shaders/rrect_blur.frag 
Mali Offline Compiler v8.4.1 (Build 215478)
Copyright (c) 2007-2024 Arm Limited. All rights reserved.

Configuration
=============

Hardware: Mali-G78 r1p1
Architecture: Valhall
Driver: r48p0-00rel0
Shader type: OpenGL ES Fragment

Compiler messages
=================

ERROR: 0:7: P0001: Unknown preprocessing directive 'include'
     
  >> #include <impeller/gaussian.glsl>
     #include <impeller/types.glsl>

ERROR: 0:8: P0001: Unknown preprocessing directive 'include'
     #include <impeller/gaussian.glsl>
  >> #include <impeller/types.glsl>
     
ERROR: 0:10: L0001: Typename expected, found 'FragInfo'
     
  >> uniform FragInfo {
       f16vec4 color;

ERROR: 0:10: L0001: Expected identifier, found '{'
     
  >> uniform FragInfo {
       f16vec4 color;

Compilation failed.
```

**Comment by gaaclarke** [comment]
For mali-t880 longest/shortest path, (14.85 + 9.0) < (25.74 + 1.0) so it still seems like a net positive.

**Comment by chinmaygarde** [comment]
If you have malioc installed locally, then it should be possible to just run the same script the bots are running to generate the diff. If malioc not in your path perhaps?

**Comment by jonahwilliams** [comment]
So I'd be careful with vectorization. Most GPUs are SIMT now, and so despite data types being vec4 based all of the math ops are still scalar. 

> My sense was that a stack spill was an instakill to performance. 

This is correct, a stack spill is absolutely catastrophic for performance.

**Comment by jonahwilliams** [comment]
I would like to spend some time verifying the reported performance numbers. I had _previously_ investigate this and on iOS found that it was net negative.

**Comment by gaaclarke** [comment]
I got rid of the stack spill by reusing the integral variable:
```
$ malioc --fragment --core Mali-T880 ../out/host_debug_unopt_arm64/gen/flutter/impeller/entity/gles/rrect_blur.frag.gles
Mali Offline Compiler v8.4.1 (Build 215478)
Copyright (c) 2007-2024 Arm Limited. All rights reserved.

Configuration
=============

Hardware: Mali-T880 r2p0
Architecture: Midgard
Driver: r23p0-00rel0
Shader type: OpenGL ES Fragment

Main shader
===========

Work registers: 7 (87% used at 50% occupancy)
Uniform registers: 3 (17% used)
Stack spilling: false

                                A      LS       T    Bound
Total instruction cycles:   13.33    1.00    0.00        A
Shortest path cycles:       12.87    1.00    0.00        A
Longest path cycles:        12.87    1.00    0.00        A

A = Arithmetic, LS = Load/Store, T = Texture

Shader properties
=================

Has uniform computation: false

Note: This tool shows only the shader-visible property state.
API configuration may also impact the value of some properties.
```

**Comment by jonahwilliams** [comment]
Numbers check out locally for me. Good to hear that you got rid of the stack spill!


 LGTM

File: impeller/compiler/shader_lib/impeller/gaussian.glsl [hunk 1]
```diff
@@ -14,6 +14,11 @@ float IPGaussian(float x, float sigma) {
   return exp(-0.5f * x * x / variance) / (kSqrtTwoPi * sigma);
 }
 
+vec4 IPGaussian(vec4 x, float sigma) {
```
Comment (jonahwilliams) [inline impeller/compiler/shader_lib/impeller/gaussian.glsl:17]: Can you document what X is? In this case, I think its x1, y2, x2, y2 right?

File: impeller/compiler/shader_lib/impeller/gaussian.glsl [hunk 2]
```diff
@@ -56,6 +61,10 @@ vec2 IPVec2FastGaussianIntegral(vec2 x, float sigma) {
   return 1.0 / (1.0 + exp(-kSqrtThree / sigma * x));
 }
 
+vec4 IPVec2FastGaussianIntegral(vec4 x, float sigma) {
```
Comment (jonahwilliams) [inline impeller/compiler/shader_lib/impeller/gaussian.glsl:64]: same here

**Comment by chinmaygarde** [comment]
> I got rid of the stack spill by reusing the integral variable

Awesome. I have no concerns past that. Seems like Jonah had some prior experience not seeing expected results with vectorization so get a +1 from him as well.

**Comment by chinmaygarde** [comment]
> Numbers check out locally for me. 

Nevermind 🙂

File: impeller/compiler/shader_lib/impeller/gaussian.glsl (hunk 1 above)
Comment (gaaclarke) [reply impeller/compiler/shader_lib/impeller/gaussian.glsl:18]: Nope, it's the same calculation just pipelined to happen 4x.  Docstring added.

File: impeller/compiler/shader_lib/impeller/gaussian.glsl (hunk 2 above)
Comment (gaaclarke) [reply impeller/compiler/shader_lib/impeller/gaussian.glsl:67]: This one is integrating over the range x to y and w to z, docstring added.

**Comment by chinmaygarde** [comment]
Hmm, the edges don't look quite right 
Screenshot 2024-10-02 at 3 40 38 PM

**Comment by gaaclarke** [comment]
The error was introduced in 56a32c7c27d0dbce1b73df7d62c45b0a89e96206 when i tried to avoid the stack spilling.  Let me double check that diff.

**Comment by gaaclarke** [comment]
fixed, I got my swizzled swizzled

**Pull request merged**